internal/pkg/deploy/pipeline.go

```diff
@@ -442,41 +445,197 @@ func (s *GitHubV1Source) Owner() (string, error) {
+	stg.execRoleARN = env.ExecutionRoleARN
+	stg.envManagerRoleARN = env.ManagerRoleARN
 }
 
-// WorkloadTemplateConfigurationPath returns the full path to the workload CFN
-// template configuration file built during the build stage.
-func (s *PipelineStage) WorkloadTemplateConfigurationPath(wlName string) string {
-	return fmt.Sprintf(WorkloadCfnTemplateConfigurationNameFormat,
-		wlName, s.Name,
-	)
+// Name returns the stage's name.
+func (stg *PipelineStage) Name() string {
+	return stg.associatedEnvironment.Name
 }
 
-// AssociatedEnvironment defines the necessary information a pipeline stage
-// needs for an Config Environment.
-type AssociatedEnvironment struct {
-	// Name of the environment, must be unique within an application.
-	// This is also the name of the pipeline stage.
-	Name string
+// Approval returns a manual approval action for the stage.
+// If the stage does not require approval, then returns nil.
+func (stg *PipelineStage) Approval() *ManualApprovalAction {
+	if !stg.requiresApproval {
+		return nil
+	}
+	return &ManualApprovalAction{
+		name: stg.associatedEnvironment.Name,
+	}
+}
 
-	// The region this environment is stored in.
-	Region string
+func (stg *PipelineStage) Region() string {
+	return stg.associatedEnvironment.Region
+}
 
-	// AccountID of the account this environment is stored in.
-	AccountID string
+// ExecRoleARN returns the IAM role assumed by CloudFormation to create or update resources defined in a template.
+func (stg *PipelineStage) ExecRoleARN() string {
+	return stg.execRoleARN
+}
+
+// EnvManagerRoleARN returns the IAM role used to create or update CloudFormation stacks in an environment.
+func (stg *PipelineStage) EnvManagerRoleARN() string {
+	return stg.envManagerRoleARN
+}
+
+// Test returns a test for the stage.
+// If the stage does not have any test commands, then returns nil.
+func (stg *PipelineStage) Test() *TestCommandsAction {
+	if len(stg.testCommands) == 0 {
+		return nil
+	}
+
+	var prevActions []orderedRunner
+	for _, deployment := range stg.Deployments() {
+		prevActions = append(prevActions, &deployment)
+	}
+
+	return &TestCommandsAction{
+		action: action{
+			prevActions: prevActions,
+		},
+		commands: stg.testCommands,
+	}
+}
+
+// Deployments returns a list of deploy actions for the pipeline.
+func (stg *PipelineStage) Deployments() []WorkloadDeployAction {
+	var prevActions []orderedRunner
+	if approval := stg.Approval(); approval != nil {
+		prevActions = append(prevActions, approval)
+	}
+
+	var actions []WorkloadDeployAction
+	for _, workload := range stg.localWorkloads {
+		actions = append(actions, WorkloadDeployAction{
+			action: action{
+				prevActions: prevActions,
+			},
+			name:    workload,
+			envName: stg.associatedEnvironment.Name,
+			appName: stg.AppName,
+		})
+	}
+	return actions
+}
+
+type orderedRunner interface {
+	RunOrder() int
+}
+
+// action represents a generic CodePipeline action.
+type action struct {
+	prevActions []orderedRunner
```

Comment:
Why do we need this 🤔 Can we record the max run order of the prev actions?

Reply:
We could definitely do that! but then that means that the client of `PipelineStage` needs to call always the methods in a specific order:
1. `Approval`
2. `Deployments`
3. `Test`

Ideally, users should be able to call `stg.Test` or `stg.Deployments` and always get the same run order. This way of implementing is just one way! We can do some caching, but my thought was that we don't expect a huge slice anyway so it's most likely fine to just do a quick for-loop

Reply:
the nitest nit: I'm not sure if this is necessary but maybe we can document this as `// The last actions to be executed immediately before this action` so that in the future we don't misunderstand this as "aggregated previous version" 🤔 ?